File: go/linkedList/25.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := new(ListNode)
	dummy.Next = head
	pre := dummy

	for head != nil {
		tail := pre
		for i := 0; i < k; i++ {
			tail = tail.Next
			if tail == nil {
				return dummy.Next
			}
		}

		next := tail.Next
		reverse(head, tail)

		pre.Next, pre, head = tail, head, next
	}
	return dummy.Next
}

func reverse(head, tail *ListNode) {
	node := tail.Next
	for node != tail {
		next := head.Next
		head.Next, node, head = node, head, next
	}
}

File: php/offer/25.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution
{

	/**
	 * @param ListNode $l1
	 * @param ListNode $l2
	 * @return ListNode
	 */
	function mergeTwoLists ($l1, $l2)
	{
		$l = new ListNode(null);
		$cur = $l;

		while ($l1 != null && $l2 != null) {
			if ($l1->val < $l2->val) {
				$cur->next = $l1;
				$l1 = $l1->next;
			} else {
				$cur->next = $l2;
				$l2 = $l2->next;
			}
			$cur = $cur->next;
		}
		if ($l1) {
			$cur->next = $l1;
		}
		if ($l2) {
			$cur->next = $l2;
		}

		return $l->next;
	}
}